### PR TITLE
enhanced xpect editor override feature

### PR DESCRIPTION
**Pull request by mehmet-karaman**
Added the ability to disable editor override at all or exclude override by extension.

### CODE REVIEW

**Comment by cdietrich** [comment]
@iloveeclipse can you please review this functionality and code whise....
wont have time to setup and env and try

**Comment by cdietrich** [comment]
thx @trancexpress @mehmet-karaman

File: org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/preferences/XpectRootPreferencePage.java [hunk 1]
```diff
@@ -12,22 +12,51 @@
...
+			
+			Composite skipFileExtListComposite = new Composite(group, SWT.NONE);
+			GridLayoutFactory.fillDefaults().margins(5, 5).applyTo(skipFileExtListComposite);
+			GridDataFactory.fillDefaults().grab(true, false).applyTo(skipFileExtListComposite);
+			addField(new SkipFileExtensionList(LIVE_TEST_EXECUTION_PREFERENCE_NAME, "Skip content check for XPECT for extensions (will not be used if override is disabled):", skipFileExtListComposite));
+			
+			Composite disableOverrideComposite = new Composite(group, SWT.NONE);
+			GridLayoutFactory.fillDefaults().margins(5, 5).applyTo(disableOverrideComposite);
+			GridDataFactory.fillDefaults().grab(true, false).applyTo(disableOverrideComposite);
+			addField(new BooleanFieldEditor(DISABLE_EDITOR_OVERRIDE_PREFERENCE_NAME, "Disable editor override.", SWT.NONE, disableOverrideComposite));
```
Comment (trancexpress) [inline org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/preferences/XpectRootPreferencePage.java:58]: Please remove the full stop symbol here, its unusual to see it in a preference page.

File: org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/preferences/XpectRootPreferencePage.java [hunk 2]
```diff
@@ -12,22 +12,51 @@
...
+		if (!isPropertyPage()) {
+			Group group = new Group(parent, SWT.NONE);
+			group.setText("Xpect Editor");
+			GridLayoutFactory.fillDefaults().margins(5, 5).applyTo(group);
+			GridDataFactory.fillDefaults().span(2, 1).grab(true, false).applyTo(group);
+			
+			Composite skipFileExtListComposite = new Composite(group, SWT.NONE);
+			GridLayoutFactory.fillDefaults().margins(5, 5).applyTo(skipFileExtListComposite);
+			GridDataFactory.fillDefaults().grab(true, false).applyTo(skipFileExtListComposite);
+			addField(new SkipFileExtensionList(LIVE_TEST_EXECUTION_PREFERENCE_NAME, "Skip content check for XPECT for extensions (will not be used if override is disabled):", skipFileExtListComposite));
```
Comment (trancexpress) [inline org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/preferences/XpectRootPreferencePage.java:53]: I think we can skip the "will not be used if override is disabled", there is an overlap of the 2 preferences but I doubt the user cares.

Maybe you can disable the entire group with the settings per preference, if the checkbox for the global disable is ticket. A nice to have though, I don't think we need much fancy UI work here.

Comment (trancexpress) [reply org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/preferences/XpectRootPreferencePage.java:53]: "Skip content check for XPECT for extensions" doesn't seem like optimal wording.

"Suppress Xpect editor override for the following file extensions"? Not ideal but maybe better.

**Comment by trancexpress** [comment]
> thx @trancexpress @mehmet-karaman

Thank you for merging Christian, but please wait a bit for reviews.

**Comment by trancexpress** [comment]
Sounds fine to me. Should be simple to reverse the logic?

**Comment by cdietrich** [comment]
you mean by extending the extension point(s)?

**Comment by iloveeclipse** [comment]
I mean the code `XpectEditorAssociationOverride.overrideDefaultEditor()` currently "detects" xpect content type also by **looking into the file itself** which means the file is always read and if it contains `XPECT` xpect assumes it should "own" the editor.

So we added a check for files that we don't want to handle with Xpect before. Fine:

https://github.com/eclipse/Xpect/blob/b37f5ddf5de67538037dd40e33362fe417c15261/org.eclipse.xpect.ui/src/org/eclipse/xpect/ui/editor/XpectEditorAssociationOverride.java#L79-L85


Instead (or additionally) we could add check if the file in the list of "known" Xpect files, because this is a smaller & fixed set, something like:

```
IFile file = getFile(editorInput);
		if (file == null || hasFavoriteEditor(file) || XpectRootPreferencePage.getSkipExtensionsList().contains(file.getFileExtension()))
			return editorDescriptor;

Set<String> knownXpectTypes = XpectRootPreferencePage.getKnownXpectExtensionsList();
if( !knownXpectTypes.isEmpty() && !knownXpectTypes.contains(file.getFileExtension())) {
			return editorDescriptor;
}
```

**Comment by cdietrich** [comment]
i am not sure what side effects that would have and what a 
"known" xpect file would mean here. so how to populates knownXpectTypes.

this is why i pointed to the extension point. that registers file extensions to xpect.

**Comment by iloveeclipse** [comment]
>  so how to populates knownXpectTypes.

Same was we populate `getSkipExtensionsList()`?

The `FileExtensionInfoRegistry` registry doesn't help here **I guess** if it works on the file extensions contributed by Xtext in the **current** IDE, as the Xtext languages in the target platform not necessarily same we have in the IDE.

The product has 10 "Xpect" file types, IDE has zero.

**Comment by mehmet-karaman** [comment]
The Extension point is not usable, because the file extensions are necessary in the IDE where the DSL (and of course its tests) is/are  developed, that's why I've introduced the new settings in the preferences page. But I am ok with the idea that we should reverse the logic. To not break anything else we should disable this by default.

**Pull request merged**